### PR TITLE
Finish createPost except image upload and location service

### PR DESCRIPTION
**Pull request by isabella0428**

### CODE REVIEW

**Comment by isabella0428** [comment]
First we have an initial page like following.
homepage

**Comment by isabella0428** [comment]
Then we can get all user names from server and select the ones we want. A click icon will exist if we select the user. Click save to return to the createPost page.
mentionUser

**Comment by isabella0428** [comment]
Then we can fill in description and select share_mode, finally click save icon to create post in database.
finalPage

File: backend/graminsta/core/services.py [hunk 1]
```diff
@@ -37,6 +37,22 @@ def create_userinfo(validated_data):
     return user_info
 
 
+def get_all_username():
```
Comment (Time1ess) [inline backend/graminsta/core/services.py]: As this happens in backend, I think you can return all `User`s instead of only the `username`s, so this API can be reused by other clients. You can pick whatever fields you need in your `views`.

File: backend/graminsta/post/services.py [hunk 2]
```diff
@@ -30,15 +30,27 @@ def create_post(publisher_id, description, img, mention_user_ids):
         publisher=publisher,
         description=description,
         img=img,
+        shared_mode=shared_mode
```
Comment (Time1ess) [inline backend/graminsta/post/services.py:34]: No parameter documentation for this?

File: backend/graminsta/post/views.py [hunk 3]
```diff
@@ -90,11 +91,33 @@ def post(self, request):
         # _ is not allowed in header key
         # TODO: get user from request
         publisher_id = int(request.META.get("HTTP_PUBLISHERID"))
+        shared_mode = request.META.get("HTTP_SHAREDMODE")
```
Comment (Time1ess) [inline backend/graminsta/post/views.py]: Hmm, why are you still retrieving informations from META? I think what you need to use is `request.data`.

File: backend/graminsta/post/views.py [hunk 4]
```diff
@@ -90,11 +91,33 @@ def post(self, request):
...
+        Returns
+        ------------
+        response: json format
+            All posts
+        """
+        posts = get_all_post()
+        body = ""
+        for post in posts:
+            body += json.dumps(PostSerializer(post).data)
+
```
Comment (Time1ess) [inline backend/graminsta/post/views.py:175]: What you need is `PostSerializer(posts, many=True`.

See also: https://www.django-rest-framework.org/api-guide/serializers/#serializing-multiple-objects

File: flutter/Graminsta/lib/main.dart [hunk 5]
```diff
@@ -10,8 +12,10 @@ class MyApp extends StatelessWidget {
   Widget build(BuildContext context) {
     return MaterialApp(
       title: "Graminsta",
-      initialRoute: '/home',
+      initialRoute: '/create',
```
Comment (Time1ess) [inline flutter/Graminsta/lib/main.dart]: Revert the change? As the initial route shouldn't be a page to create post?

File: flutter/Graminsta/lib/post/createPostWidget.dart [hunk 6]
```diff
@@ -48,7 +59,25 @@ class _CreatePostWidgetState extends State<CreatePostWidget> {
                 Icons.save,
                 color: Colors.white,
               ),
-              onPressed: () {})
+              onPressed: () async {
+                final http.Response response = await http.post(
+                    "http://10.0.2.2:8000/post/",
```
Comment (Time1ess) [inline flutter/Graminsta/lib/post/createPostWidget.dart]: Do not submit local ip

File: flutter/Graminsta/lib/post/createPostWidget.dart [hunk 7]
```diff
@@ -48,7 +59,25 @@ class _CreatePostWidgetState extends State<CreatePostWidget> {
...
+                    headers: <String, String>{
+                      "PublisherId": "1",
+                      "SharedMode": accessControl.toHumanReadableString(),
+                    },
```
Comment (Time1ess) [inline flutter/Graminsta/lib/post/createPostWidget.dart]: Put your business data into `body` instead of `headers`.

File: flutter/Graminsta/lib/post/createPostWidget.dart [hunk 8]
```diff
@@ -48,7 +59,25 @@ class _CreatePostWidgetState extends State<CreatePostWidget> {
...
+                    body: {
+                      "description": myController.text,
+                      "mention_usernames": mentionedUser,
+                      "img": "this is my img",
+                    });
+
+                if (response.statusCode == 201) {
+                  print(response.body);
```
Comment (Time1ess) [inline flutter/Graminsta/lib/post/createPostWidget.dart]: use logging module to print your debug information. `print` almost means nothing in a Flutter app.

File: flutter/Graminsta/lib/post/createPostWidget.dart [hunk 9]
```diff
@@ -86,11 +116,22 @@ class _CreatePostWidgetState extends State<CreatePostWidget> {
...
+                        onPressed: () async {
+                          String result = await Navigator.push(
+                              context,
+                              MaterialPageRoute(
+                                  builder: (context) => MentionUserWidget()));
+                          setState(() {
+                            print(mentionedUser);
+                            mentionedUser = result.replaceAll('"', '');
+                          });
+                        })),
```
Comment (Time1ess) [inline flutter/Graminsta/lib/post/createPostWidget.dart]: Add comma

File: flutter/Graminsta/lib/post/createPostWidget.dart [hunk 10]
```diff
@@ -86,11 +116,22 @@ class _CreatePostWidgetState extends State<CreatePostWidget> {
...
+                    trailing: IconButton(
+                        icon: Icon(Icons.arrow_right, size: 30),
+                        iconSize: 48,
+                        onPressed: () async {
+                          String result = await Navigator.push(
+                              context,
+                              MaterialPageRoute(
+                                  builder: (context) => MentionUserWidget()));
+                          setState(() {
+                            print(mentionedUser);
```
Comment (Time1ess) [inline flutter/Graminsta/lib/post/createPostWidget.dart]: no print

File: flutter/Graminsta/lib/post/createPostWidget.dart [hunk 11]
```diff
@@ -86,11 +116,22 @@ class _CreatePostWidgetState extends State<CreatePostWidget> {
                   trailing: Icon(Icons.arrow_right, size: 30),
                 ),
                 new ListTile(
-                  leading: const Icon(Icons.alternate_email),
-                  title: Text("Mention"),
-                  trailing: Icon(Icons.arrow_right, size: 30),
-                ),
-                ListTile(
+                    leading: const Icon(Icons.alternate_email),
+                    title: Text('Mention:                 ' + mentionedUser),
```
Comment (Time1ess) [inline flutter/Graminsta/lib/post/createPostWidget.dart]: Can we avoid so many hard-coded white spaces in the text?

File: flutter/Graminsta/lib/post/mentionUserWidget.dart [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+class MentionUserWidget extends StatefulWidget {
+  MentionUserWidget({Key key}) : super(key: key);
+
+  @override
+  _MentionUserWidgetState createState() => _MentionUserWidgetState();
+}
+
+class _MentionUserWidgetState extends State<MentionUserWidget> {
+  List<IconData> userIcons = new List();
+  List<String> allNames;
```
Comment (Time1ess) [inline flutter/Graminsta/lib/post/mentionUserWidget.dart:15]: As I suggested before. You should display the name (not username) of users that have been selected to the user. And you should use the id of the selected user in your business logic. You should never expose `username`.

File: flutter/Graminsta/pubspec.lock [hunk 13]
```diff
@@ -1,55 +1,83 @@
...
+      name: args
+      url: "https://pub.dartlang.org"
+    source: hosted
+    version: "1.6.0"
   async:
     dependency: transitive
     description:
       name: async
-      url: "https://mirrors.tuna.tsinghua.edu.cn/dart-pub"
+      url: "https://pub.dartlang.org"
```
Comment (Time1ess) [inline flutter/Graminsta/pubspec.lock]: Please set and keep using tsinghua mirror, otherwise it's too slow to download packages from dartlang.org.

**Review by Time1ess — CHANGES_REQUESTED** [review]

File: flutter/Graminsta/lib/post/createPostWidget.dart (hunk 6 above)
Comment (isabella0428) [reply flutter/Graminsta/lib/post/createPostWidget.dart]: It is the default localhost for android simulator

File: flutter/Graminsta/lib/post/mentionUserWidget.dart (hunk 12 above)
Comment (isabella0428) [reply flutter/Graminsta/lib/post/mentionUserWidget.dart:15]: You mean first name and last name? But these fields are optional.

File: flutter/Graminsta/lib/post/createPostWidget.dart (hunk 6 above)
Comment (Time1ess) [reply flutter/Graminsta/lib/post/createPostWidget.dart]: I mean, this config is only for local development, it's not for production use, so you'd better to have some `settings_dev.dart`, `settings_prod.dart` for different envs.

And a `settings.dart` to export different settings for different envs (you can manually export in that file if conditional export is not easy to achieve), and all your code should reference settings from `settings.dart`

File: flutter/Graminsta/lib/post/createPostWidget.dart (hunk 9 above)
Comment (Time1ess) [reply flutter/Graminsta/lib/post/createPostWidget.dart]: Not resolved? Add comma before the last `)`

File: flutter/Graminsta/lib/post/mentionUserWidget.dart (hunk 12 above)
Comment (Time1ess) [reply flutter/Graminsta/lib/post/mentionUserWidget.dart:15]: Yes. You should assume that the first name and last name are populated, other parts will take care of that, even though in our dev env these fields are not correctly populated.

first_name, last_name --> displayed in the UI
id of the user --> sent to server

I can't think of a case where username should be used anywhere in the frontend (our Flutter app)

File: flutter/Graminsta/pubspec.yaml [hunk 14]
```diff
@@ -23,6 +23,8 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
+  http: 0.12.2
+  dio: ^3.0.10
```
Comment (Time1ess) [inline flutter/Graminsta/pubspec.yaml]: Why do we need this? since we already have `http`

**Review by Time1ess — CHANGES_REQUESTED** [review]

File: flutter/Graminsta/lib/post/mentionUserWidget.dart (hunk 12 above)
Comment (isabella0428) [reply flutter/Graminsta/lib/post/mentionUserWidget.dart:15]: Since we will not need to send to server here, I just modify username to "first_name last_name"

File: backend/graminsta/core/views.py [hunk 15]
```diff
@@ -47,28 +47,17 @@ def get(self, request):
...
         response: json format
-            username of all users that satisfy the requirements
+            displayed name of all users that satisfy the requirements
         """
-        username = get_all_username()
+        all_users = get_all_users()
+        all_users_name = []
+        separator = ","
+        for user in all_users:
+            all_users_name.append(user.first_name + " " + user.last_name)
```
Comment (Time1ess) [inline backend/graminsta/core/views.py:57]: Why not just use `UserInfoSerializer` to serialize these `User` objects. If has all the information you need

**Review by Time1ess — APPROVED** [review]

File: backend/graminsta/core/views.py (hunk 15 above)
Comment (isabella0428) [reply backend/graminsta/core/views.py:57]: The front-end only need displayed username, so i think maybe we should just return it instead of all user objects?

**Review by ArianaGu — APPROVED** [review]
